common/app/model/content/Atoms.scala

```diff
@@ -47,35 +47,47 @@ final case class Quiz(
   revealAtEnd: Boolean
 ) extends Atom
 
-
-
+final case class InteractiveAtom(
+  override val id: String,
+  `type`: String,
+  title: String,
+  css: String,
+  html: String,
+  mainJS: Option[String],
+  docData: Option[String]
+) extends Atom
 
 object Atoms extends common.Logging {
+  def extract[T](atoms: Option[Seq[AtomApiAtom]], extractFn: AtomApiAtom => T): Seq[T] = {
+    try {
+      atoms.getOrElse(Nil).map(extractFn)
+    } catch {
+      case e: Exception =>
+        logException(e)
+        Nil
+    }
+  }
+
   def make(content: contentapi.Content): Option[Atoms] = {
     content.atoms.map { atoms =>
-      val quizzes: Seq[atomapi.quiz.QuizAtom] = try {
-        atoms.quizzes.getOrElse(Nil).map(atom => {
-          atom.data.asInstanceOf[AtomData.Quiz].quiz
-        })
-      } catch {
-        case e: Exception =>
-          logException(e)
-          Nil
-      }
-      val media: Seq[MediaAtom] = try {
-        atoms.media.getOrElse(Nil).map(atom => {
-          val id = atom.id
-          val mediaAtom = atom.data.asInstanceOf[AtomData.Media].media
-          MediaAtom.mediaAtomMake(id, mediaAtom)
-        })
-      } catch {
-        case e: Exception =>
-          logException(e)
-          Nil
-      }
+      val quizzes = extract(atoms.quizzes, atom => {
+        val quizAtom = atom.data.asInstanceOf[AtomData.Quiz].quiz
+        Quiz.make(content.id, quizAtom)
+      })
+
+      val media = extract(atoms.media, atom => {
+        val id = atom.id
+        val mediaAtom = atom.data.asInstanceOf[AtomData.Media].media
+        MediaAtom.mediaAtomMake(id, mediaAtom)
+      })
+
+      val interactives = extract(atoms.interactives, atom => {
+        val interactiveAtom = atom.data.asInstanceOf[AtomData.Interactive].interactive
+        InteractiveAtom.make(atom.id, interactiveAtom)
+      })
```

Comment:
What about passing the ApiAtom and matching using `isInstanceOf` as well as using `collect` rather than `map`

Like this
```
def make(content: contentapi.Content): Option[Atoms] = {

    val apiAtomToQuiz: PartialFunction[AtomApiAtom, Quiz] = {
      case atom if atom.data.isInstanceOf[AtomData.Quiz] => Quiz.make(content.id, atom.data.asInstanceOf[AtomData.Quiz].quiz)
    }
    val apiAtomToMedia: PartialFunction[AtomApiAtom, MediaAtom] = {
      case atom if atom.data.isInstanceOf[AtomData.Media] => MediaAtom.mediaAtomMake(content.id, atom.data.asInstanceOf[AtomData.Media].media)
    }
    val apiAtomToInteractive: PartialFunction[AtomApiAtom, InteractiveAtom] = {
      case atom if atom.data.isInstanceOf[AtomData.Interactive] => InteractiveAtom.make(atom.id, atom.data.asInstanceOf[AtomData.Interactive].interactive)
    }

    def extractAtoms[T](atoms: Option[Seq[AtomApiAtom]], apiAtomtoAtom: PartialFunction[AtomApiAtom, T]): Seq[T] = atoms.fold(Seq.empty[T])(_.collect(apiAtomtoAtom))

    content.atoms.map { atoms =>
      Atoms(
        quizzes = extractAtoms(atoms.quizzes, apiAtomToQuiz),
        media = extractAtoms(atoms.media, apiAtomToMedia),
        interactives = extractAtoms(atoms.quizzes, apiAtomToInteractive)
      )
    }
  }
```